packages/netsuite-adapter/src/filters/instance_references.ts

```diff
@@ -127,34 +139,47 @@ const replaceReferenceValues = async (
   fetchedElementsServiceIdToElemID: Record<string, ElemID>,
   elementsSourceServiceIdToElemID: Record<string, ElemID>,
 ): Promise<InstanceElement> => {
+  const dependenciesToAdd: Array<ElemID> = []
   const replacePrimitive: TransformFunc = ({ path, value }) => {
     if (!_.isString(value)) {
       return value
     }
     const serviceIdInfo = captureServiceIdInfo(value)
-    if (_.isUndefined(serviceIdInfo)) {
-      return value
-    }
-    const elemID = fetchedElementsServiceIdToElemID[serviceIdInfo[CAPTURED_SERVICE_ID]]
-      ?? elementsSourceServiceIdToElemID[serviceIdInfo[CAPTURED_SERVICE_ID]]
-
-    const type = serviceIdInfo[CAPTURED_TYPE]
-    if (_.isUndefined(elemID) || (type && type !== elemID.typeName)) {
-      return value
-    }
-
-    if (path?.isAttrID() && path.createParentID().name === CORE_ANNOTATIONS.PARENT) {
-      return new ReferenceExpression(elemID.createBaseID().parent)
-    }
+    let returnValue: ReferenceExpression | string = value
+    serviceIdInfo.forEach(serviceIdInfoRecord => {
+      const elemID = fetchedElementsServiceIdToElemID[serviceIdInfoRecord[CAPTURED_SERVICE_ID]]
+      ?? elementsSourceServiceIdToElemID[serviceIdInfoRecord[CAPTURED_SERVICE_ID]]
+
+      const type = serviceIdInfoRecord[CAPTURED_TYPE]
+      if (_.isUndefined(elemID) || (type && type !== elemID.typeName)) {
+        return
+      }
+
+      if (path?.isAttrID() && path.createParentID().name === CORE_ANNOTATIONS.PARENT) {
+        if (serviceIdInfoRecord.isFullMatch) {
+          returnValue = new ReferenceExpression(elemID.createBaseID().parent)
+          return
+        }
+        dependenciesToAdd.push(elemID.createBaseID().parent)
+        return
+      }
+      if (serviceIdInfoRecord.isFullMatch) {
+        returnValue = new ReferenceExpression(elemID)
+        return
+      }
+      dependenciesToAdd.push(elemID)
+    })
 
-    return new ReferenceExpression(elemID)
+    return returnValue
   }
 
-  return transformElement({
+  const newInstance = await transformElement({
     element: instance,
     transformFunc: replacePrimitive,
     strict: false,
   })
+  dependenciesToAdd.map(dep => addElementGeneratedDepRef(newInstance, dep))
```

Comment:
I think that something like the below suggestions, which are generally the same, would be clearer. If you choose my suggestion, please extract it to a method :)
This is just a matter of styling and your impl is correct.
Although your impl is less sufficient, it doesn't really matter because we don't expect to have mass of such elements...

```suggestion
newInstance.annotations[CORE_ANNOTATIONS.GENERATED_DEPENDENCIES] = _.uniqBy([...makeArray(newInstance.annotations[CORE_ANNOTATIONS.GENERATED_DEPENDENCIES]), ... dependenciesToAdd], ref => ref.elemID.getFullName())

newInstance.annotations[CORE_ANNOTATIONS.GENERATED_DEPENDENCIES] = _.uniqBy(makeArray(newInstance.annotations[CORE_ANNOTATIONS.GENERATED_DEPENDENCIES]).concat(dependenciesToAdd), ref => ref.elemID.getFullName())
```